### PR TITLE
feat(x/ecocredit/v3/client): add/update independent project CLI commands

### PR DESCRIPTION
**Pull request by aaronc**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by aaronc** [comment]
Merged this into #2167

File: x/ecocredit/base/client/query.go [hunk 1]
```diff
@@ -667,3 +667,66 @@ func QueryAllowedBridgeChainsCmd() *cobra.Command {
 	flags.AddPaginationFlagsToCmd(cmd, "allowed-bridge-chains")
 	return qflags(cmd)
 }
+
+func QueryProjectEnrollment() *cobra.Command {
```
Comment (paul121) [inline x/ecocredit/base/client/query.go:671]: consistency
```suggestion
func QueryProjectEnrollmentCmd() *cobra.Command {

```

File: x/ecocredit/base/client/query.go [hunk 2]
```diff
@@ -667,3 +667,66 @@ func QueryAllowedBridgeChainsCmd() *cobra.Command {
...
+				ProjectId: args[0],
+				ClassId:   args[1],
+			})
+			return printQueryResponse(ctx, res, err)
+		},
+	}
+	return qflags(cmd)
+}
+
+func QueryProjectEnrollments() *cobra.Command {
```
Comment (paul121) [inline x/ecocredit/base/client/query.go:693]: consistency

```suggestion
func QueryProjectEnrollmentsCmd() *cobra.Command {

```

**Review by paul121 — COMMENTED** [review]
This is working well! I didn't find any major state or logic issues while testing. Tests seem to cover this pretty well :+1:

Just found various little things, mostly that a few of the CLI commands weren't quite wired up and could maybe be improved for consistency.

Simple fixes I can't suggest via github UI:
- Need to add project enrollment query cmds to x/ecocredit/client/query.go QueryCmd():

    ```
            baseclient.QueryProjectEnrollmentCmd(),
            baseclient.QueryProjectEnrollmentsCmd(),
    ```

- Error with the example batch.json in `create-batch` command help text. It is missing a `class_id`
- Doesn't need to be fixed in this PR, but the `ecocredit` script in images/regen-sandbox/setup/ecocredit.sh needs some updating (some commands fail due to these changes)

The general cases I tested:
- Creating projects independent vs direct, testing correct permissions to update project metadata, etc
- Creating application, withdrawing application, testing correct permissions + good/bad project + class IDs, metadata
- Updating enrollment status, testing correct permissions, correct status transitions, good/bad project + class IDs, metadata
- Creating credit batch, testing correct permissions, good/bad project + class IDs
- "Full flow": Create project, apply, enroll, issue credits, terminate enrollment, try to issue credits, re-enroll, issue credits

I think this covered using/testing the following CLI commands:

- tx ecocredit:
  - create-class
  - create-project 
  - update-project-metadata
  - create-or-update-application
  - withdraw-application
  - update-project-enrollment
  - create-batch
  - gen-batch-json

- q ecocredit:
  - batches
  - class, classes
  - class-issuers
  - project, projects
  - project-enrollment
  - project-enrollments


Could improve error messages when proj + class enrollments do not exist/have an entry in the enrollment DB table:
- withdraw-application
  - when the credit class or project ID does not exist gets generic error msg: `raw_log: 'failed to execute message; message index: 0: not found'`
  - when valid credit class and project ID, but no application/enrollment between the two, gets a better error msg: `raw_log: 'failed to execute message; message index: 0: cannot withdraw non-existent application: invalid request'`
- create-batch
  - when credit class does not exist or when no enrollment for project + class: `raw_log: 'failed to execute message; message index: 0: not found'`
  - when project does not exist, better error msg: `raw_log: 'failed to execute message; message index: 0: could not get project with id P005: not found: invalid request'`

Project enrollment status - just an idea
- There is no way for a credit class to signal that they have "received" an application and are beginning review, maybe give an expected timeframe, etc. I assume this could happen via enrollment metadata.
  - The credit class can only add metadata when setting the status to ACCEPTED or CHANGES_REQUESTED.
  - Would it make sense to add a project enrollment status for "PENDING" / "IN REVIEW" so there is an intermediary status?

**Pull request closed without merging**